src/runtime_src/doc/toc/xrt_native_apis.rst

```diff
@@ -786,3 +786,110 @@ The above code shows
+  - The task enqueued on the queue must be synchronous in nature
+  - All the tasks enqueued on the queue will be completed in the order it is enqueued (strict in-order execution). 
+  - The task enqueued on a queue can be any C++ Callable, which can conveniently be expressed by a C++ lambda
+  - When a synchronous task is enqueued on a queue, an event (``xrt::queue:event``) is returned. The event can be used for multiple purposes, such as: 
+  
+       - The host-thread can wait on that event to synchronize with the ``queue::enqueue(task)`` from where the event was generated. 
+       - The event can be enqueued to other queues to synchronize among the queues 
+       
+Let's understand the above properties by reviewing the examples below
+
+Executing synchronous task asynchronously 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+.. code:: c++
+      :number-lines: 41
+      
+          auto bo0 = xrt::bo(device, vector_size_bytes, krnl.group_id(0));
+          auto bo0_map = bo0.map<dtype*>();
+          .... // fill buffer content
+
+          xrt::queue my_queue;
+          auto sync_event = queue.enqueue([&bo0] {bo0.sync(XCL_BO_SYNC_BO_TO_DEVICE); });
+
+          myCpuTask(b);  // here we can perform other host task that will run parallel to the above bo::sync task
+          
+          sync_event.wait();  // stall the host-thread till sync operation completes 
+          
+The above code shows the synchronous API ``xrt::bo::sync`` is enqueued through an ``xrt::queue``. The argument of xrt::queue is unnamed callable written using C++ lambda capturing buffer object. This technique is useful to execute any synchronous task asynchronously from the host-thread, and while this task is ongoing, the host-thread can do other operation in parallel (``myCpuTask()`` in the above code). The return type of ``xrt::queue::enqueue()`` is type of ``xrt::queue::event`` which is later synchronized to the host-thread by ``xrt::queue::event::wait()`` blocking function. 
+
+
+Executing multiple tasks through queue 
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Every new ``xrt::queue`` can be thought of as a new thread running parallel to the host-thread executing a series of synchronous tasks following the order they were submitted (enqueued) on the queue. For example, let's consider tasks A, B, C and D as below
+   
+    - Task A: Host to device data transfer (buffer bo0)
+    - Task B: Execute the kernel and wait for the kernel to finish execution
+    - Task C: Device to host data transfer (buffer bo_out) 
+    - Task D: Check return data bo_out
+    
+The above four tasks should be executed in-order for correct functionality. To execute them in parallel to the host-thread, these four tasks can be enqueued through a queue as below. 
+
+
+.. code:: c++
+      :number-lines: 41
+      
+        
+          xrt::queue queue;
+          queue.enqueue([&bo0] {bo0.sync(XCL_BO_SYNC_BO_TO_DEVICE); });
+          queue.enqueue([&run] {run.start(); run.wait(); });
```

Comment:
Perhaps you could offer some chaining in the API like
```suggestion
          queue.enqueue([&run] {run.start().wait(); });
```

Reply:
I like it, but it requires changing the xrt::run APIs. Not sure if this can be done without breaking ABI?

Reply:
Do you have an ABI problem too? Is this API already in an official shipped release?

Reply:
This API?   Yeah, xrt::kernel and xrt::run are not new.